### PR TITLE
feat(portex): add support for object unpack grammar of Portex

### PR DESCRIPTION
**Pull request by Lee-000**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-2h0zej3 Portex schema 支持字典解包语法](https://app.clickup.com/t/2h0zej3)

File: graviti/portex/param.py [hunk 1]
```diff
@@ -119,7 +119,7 @@ def check(self, arg: Any) -> Any:
 
         """
         if arg == self.default:
-            return arg
+            return self.ptype.load(arg)  # type: ignore[union-attr]
```
Comment (linjiX) [inline graviti/portex/param.py:122]: what if the default value can not be loaded?

File: graviti/portex/factory.py [hunk 2]
```diff
@@ -249,6 +267,7 @@ def __call__(self, **kwargs: Any) -> PortexType:
 
         """
         type_kwargs = {key: factory(**kwargs) for key, factory in self._factories.items()}
+        type_kwargs.update(kwargs.get(self._unpack, {}))
```
Comment (linjiX) [inline graviti/portex/factory.py:270]: Does that mean the unpacked param as higher priority than other param?

File: graviti/portex/factory.py [hunk 3]
```diff
@@ -495,19 +540,19 @@ def __call__(self, **kwargs: Any) -> Fields:
...
     Arguments:
         decl: A dict which indicates a portex type or a dynamic portex type.
         imports: The :class:`Imports` instance to specify the import scope of the template.
 
     Returns:
-        A ``TypeFactory`` or a ``DynamicTypeFactory`` instance according to the input.
+        A ``TypeFactory`` or a ``DictUnpackTypeFactory`` instance according to the input.
 
     """
-    if decl["type"].startswith("$params."):
```
Comment (linjiX) [inline graviti/portex/factory.py:509]: which means the DynamicType is blocked after this commit?
should we block it after implementing all unpack grammars?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/portex/ptype.py [hunk 4]
```diff
@@ -196,7 +232,9 @@ def check(arg: TypingAny) -> Tuple[str, ClassPortexType]:
         raise TypeError("Argument should be a tuple of a str and a PortexType")
 
     @staticmethod
-    def load(content: Dict[str, TypingAny], imports: Imports) -> Tuple[str, ClassPortexType]:
+    def load(
+        content: Dict[str, TypingAny], imports: Optional[Imports] = None
```
Comment (linjiX) [inline graviti/portex/ptype.py:236]: Why we need to change the `imports` to `Optional`?

File: graviti/portex/factory.py [hunk 5]
```diff
@@ -495,19 +540,19 @@ def __call__(self, **kwargs: Any) -> Fields:
...
-        return DynamicTypeFactory(decl, imports)
+    if "type" not in decl and "+" in decl:
```
Comment (linjiX) [inline graviti/portex/factory.py:554]: what if there are no "$params" after "+"?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/portex/factory.py [hunk 6]
```diff
@@ -422,6 +477,7 @@ class FieldFactory(Factory):
     def __init__(self, decl: Dict[str, Any], imports: Imports) -> None:
         self.creator: Callable[..., Tuple[str, PortexType]]
 
+        print("FieldFactory decl", decl)
```
Comment (linjiX) [inline graviti/portex/factory.py:480]: remove the debug code

File: graviti/portex/factory.py [hunk 7]
```diff
@@ -455,7 +511,10 @@ def __call__(self, **kwargs: Any) -> Optional[Tuple[str, PortexType]]:
         if not self._expression(**kwargs):
             return None
 
-        return self._name_factory(**kwargs), self._type_factory(**kwargs)
+        return (
+            self._name_factory(**kwargs),
+            self._type_factory(**kwargs),  # type: ignore[return-value]
```
Comment (linjiX) [inline graviti/portex/factory.py:516]: where is the ignore from?

File: graviti/portex/ptype.py [hunk 8]
```diff
@@ -172,6 +172,42 @@ def copy(arg: _T) -> _T:
         return copy(arg)
 
 
+class ParamDict(ParameterType):
```
Comment (linjiX) [inline graviti/portex/ptype.py:175]: can use just use `Dict` or `Mapping` rather than `ParamDict`?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/portex/factory.py [hunk 9]
```diff
@@ -284,6 +302,37 @@ def __call__(self, **kwargs: Any) -> PortexType:
...
+        """Apply the input arguments to the type template.
+
+        Arguments:
+            kwargs: The input arguments.
+
+        Returns:
+            The applied Portex type.
+
+        """
+        return kwargs.get(self._key, {})
```
Comment (linjiX) [inline graviti/portex/factory.py:333]: why using `get` here, can we use `__getitem__` instead?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/portex/factory.py [hunk 10]
```diff
@@ -284,6 +302,37 @@ def __call__(self, **kwargs: Any) -> PortexType:
         return TypeFactory(decl, self._imports)(**kwargs)
 
 
+class DictUnpackFactory(Factory):
```
Comment (linjiX) [inline graviti/portex/factory.py:305]: do we really need a `DictUnpackFactory`, it looks like a `VariableFactory`

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/portex/factory.py [hunk 11]
```diff
@@ -455,7 +485,10 @@ def __call__(self, **kwargs: Any) -> Optional[Tuple[str, PortexType]]:
         if not self._expression(**kwargs):
             return None
 
-        return self._name_factory(**kwargs), self._type_factory(**kwargs)
+        return (
+            self._name_factory(**kwargs),
+            self._type_factory(**kwargs),
+        )
```
Comment (linjiX) [inline graviti/portex/factory.py:491]: seems nothing changes here?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/portex/factory.py [hunk 12]
```diff
@@ -493,19 +524,43 @@ def __call__(self, **kwargs: Any) -> Fields:
         )
 
 
+def object_unpack_factory_creator(decl: str, is_ptype: bool = False) -> VariableFactory:
```
Comment (linjiX) [inline graviti/portex/factory.py:527]: will mapping `better` than `object`

File: graviti/portex/factory.py [hunk 13]
```diff
@@ -493,19 +524,43 @@ def __call__(self, **kwargs: Any) -> Fields:
...
+        is_ptype: Whether the object unpack indicates a Portex type.
+
+    Raises:
+        ValueError: When the object unpack grammar is incorrect.
+
+    Returns:
+        A ``VariableFactory`` instance or None according to the input.
+
+    """
+    if not decl.startswith("$params."):
```
Comment (linjiX) [inline graviti/portex/factory.py:541]: is that possible to put a real mapping after `"+"` rather than use "$params."?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/portex/factory.py (hunk 12 above)
Comment (linjiX) [reply graviti/portex/factory.py:527]: Also will use `ptype` directly rather than `bool`?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/portex/factory.py (hunk 13 above)
Comment (Lee-000) [reply graviti/portex/factory.py:541]: I don't see that necessary.

File: graviti/portex/factory.py [hunk 14]
```diff
@@ -388,6 +409,12 @@ def __init__(self, decl: Dict[str, Any], ptype: PTYPE.PType = PTYPE.Any) -> None
         dependences = set()
         keys = {}
         for key, value in decl.items():
+            if key == "+":
```
Comment (linjiX) [inline graviti/portex/factory.py:412]: will it be better to check `“+”` first? rather than check it in loop?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/portex/factory.py [hunk 15]
```diff
@@ -387,6 +408,13 @@ def __init__(self, decl: Dict[str, Any], ptype: PTYPE.PType = PTYPE.Any) -> None
         factories = {}
         dependences = set()
         keys = {}
+
+        unpack_value = decl.pop("+")
```
Comment (linjiX) [inline graviti/portex/factory.py:412]: will `pop` effect input value `decl`?

**Review by linjiX — CHANGES_REQUESTED** [review]

**Review by linjiX — APPROVED** [review]

**Pull request merged**